WalletWasabi/WabiSabi/Models/Decomposition/BaseDenomGenerator.cs

```diff
@@ -0,0 +1,39 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.WabiSabi.Models.Decomposition
+{
+	public static class BaseDenominationGenerator
+	{
+		private static Money MaxSats = Money.Satoshis(Constants.MaximumNumberOfSatoshis);
+
+		private static IEnumerable<Money> Multiple(IEnumerable<long> coefficients, IEnumerable<long> values) =>
+			values.SelectMany(v => coefficients.Select(c => Money.Satoshis(c * v))).Where(x => x <= MaxSats);
+
+		private static IEnumerable<Money> Generate()
+		{
+			var powersOfTwo = Enumerable.Range(0, 50).Select(x => (long)1m << x);
+			var powersOfThree = Enumerable.Range(0, 32).Select(x => (long)Math.Pow(3, x));
+			var powersOfTen = Enumerable.Range(0, 16).Select(x => (long)Math.Pow(10, x));
+
+			var ternary = Multiple(new long[] { 1, 2 }, powersOfThree);
+			var preferredValueSeries = Multiple(new long[] { 1, 2, 5 }, powersOfTen);
+
+			return powersOfTwo
+				.Select(x => Money.Satoshis(x))
+				.Union(ternary)
+				.Union(preferredValueSeries)
+				.OrderBy(v => v);
+		}
+
+		public static IEnumerable<Money> GenerateWithEffectiveCost(FeeRate feeRate)
+		{
+			var outputSize = Constants.OutputSizeInBytes;
+			var fee = feeRate.GetFee(outputSize);
```

Comment:
We cannot know the fee without knowing the size first and we cannot know the size without knowing the scriptPubKey type. Sorry for going back to this topic again but it is my concern is not about payments in coinjoins or any other future feature but more about things like taproot scripts for example. IMO we should not use hardcoded values for things like this.

Reply:
Let me be more clear. I would remove this method. Other component, lets call it "TheFinalSuperSmartAmountDecomposer" can use your generator to generate the denominations, after that it can use the feerate information and script types for inputs and outputs to calculate the effective amount and use my decomposer to decompose the effective amount in those denominations and take care of the dust threshold and so on.

Reply:
I don't think there is a problem with hard coding a size for now. Non p2wpkh outputs are not on the agenda any time soon soon, and I promise you will have to deal with more walls of text if I think we're missing something when we get around to it.

Reply:
I removed the controversial part.

Reply:
I will add this in another PR and incrementally reach the consensus.